server/manager/src/main/java/org/apache/accumulo/manager/tableOps/delete/DeleteTable.java

```diff
@@ -43,8 +43,13 @@ public DeleteTable(NamespaceId namespaceId, TableId tableId) {
   @Override
   public long isReady(long tid, Manager env) throws Exception {
 
-    // Before attempting to delete the table, cancel any running user
-    // compactions.
+    // If we have the table write lock, then it's possible that we have
+    // already canceled the compactions and we can skip to deleting the tables.
+    if (Utils.hasNamespaceReadLock(env, namespaceId, tid)
+        && Utils.hasTableWriteLock(env, tableId, tid)) {
+      return 0;
+    }
+    // Cancel any running user compactions before deleting table
     if (Utils.reserveNamespace(env, namespaceId, tid, false, true, TableOperation.COMPACT_CANCEL)
```

Comment:
I just noticed that this will not wait when it can not get the read lock.  Another possible way to solve this problem is to add two new Repos that run before DeleteTable.  When isReady() returns a non-zero value, then the operation will be retired again later.

```java
class PredeleteTable extends ManagerRepo {
     public long isReady(long tid, Manager env) {
           // attempt to get read lock to for cancel compactoin
           return Utils.reserveNamespace(env, namespaceId, tid, false, true, TableOperation.COMPACT_CANCEL) +
               Utils.reserveTable(env, tableId, tid, false, true, TableOperation.COMPACT_CANCEL);
     }

    public Repo<Manager> call(long tid, Manager env) {
           CancelCompactions.mutateZooKeeper(tid, tableId, env);
           return PredeleteTableUnlock(...);
     }

  @Override
  public void undo(long tid, Manager env) {
           Utils.unreserveTable(env, tableId, tid, false);
           Utils.unreserveNamespace(env, namespaceId, tid, false);
  }
}

class PredeleteTableUnlock extends ManagerRepo {
     public long isReady(long tid, Manager env) {
           // attempt to get read lock to for cancel compactoin
           return 0;
     }

    public Repo<Manager> call(long tid, Manager env) {
           Utils.unreserveTable(env, tableId, tid, false);
           Utils.unreserveNamespace(env, namespaceId, tid, false);
           return DeleteTable(...);  // call the existing first fate op
     }
}
```

Also, I think this pattern may not require the hasLockMethods.  It also avoid problem w/ undo on delete table if the read lock is held and not the write lock.

Reply:
Would also need to find the code that seeds the delete table fate op w/ the DeleteTable repo and make it use PredeleteTable instead.

Reply:
I don't think you need the separate PredeleteTableUnlock, though. Couldn't you just unlock in the call method before returning the DeleteTable repo?

Reply:
> Couldn't you just unlock in the call method before returning the DeleteTable repo?

Maybe.  I was thinking about that and was not sure if it was idempotent, would need to look at how the FATE ops are run.  I was more certain that breaking it into two steps was idempotent.

Reply:
> Couldn't you just unlock in the call method before returning the DeleteTable repo?

To be clear, you are suggesting that we move the CancelCompactions.mutateZooKeeper() from it's current location in DeleteTable.isReady() to DeleteTable.call()? This seems like a much simpler change. I think I'm going to hold off from doing anything more until Keith looks at the FATE ops.

Reply:
> > Couldn't you just unlock in the call method before returning the DeleteTable repo?
> 
> To be clear, you are suggesting that we move the CancelCompactions.mutateZooKeeper() from it's current location in DeleteTable.isReady() to DeleteTable.call()?

No.

I was suggesting that instead of Keith's suggestion to have a whole PredeleteTableUnlock op just to unreserve the locks reserved in PredeleteTable.isReady, that his suggested PredeleteTable.call could do it itself before it returns `new DeleteTable(...)`

Reply:
@ctubbsii I think you were proposing the following?   I put some comments about what I was uncertain about.  If those uncertainties were resolved, I think it would be ok to have one repo prior to delete.  I suspect isReady() is called again, in that situation but I Am not certain.


```java
class PredeleteTable extends ManagerRepo {
     public long isReady(long tid, Manager env) {
           // attempt to get read lock to for cancel compactoin
           return Utils.reserveNamespace(env, namespaceId, tid, false, true, TableOperation.COMPACT_CANCEL) +
               Utils.reserveTable(env, tableId, tid, false, true, TableOperation.COMPACT_CANCEL);
     }

    public Repo<Manager> call(long tid, Manager env) {
           CancelCompactions.mutateZooKeeper(tid, tableId, env);
           Utils.unreserveTable(env, tableId, tid, false);
             // if the manager dies here, will it call isReady() again later before this method and if so, is that ok?   If isReady were not called again, could run CancelCompactions.mutateZooKeeper w/o holding lock
           Utils.unreserveNamespace(env, namespaceId, tid, false);
             // if the manager dies here, will it call isReady() again later before this method and if so, is that ok?   If isReady were not called again, could run CancelCompactions.mutateZooKeeper w/o holding lock
           return DeleteTable(...);  // call the existing first fate op
     }

  @Override
  public void undo(long tid, Manager env) {
           Utils.unreserveTable(env, tableId, tid, false);
           Utils.unreserveNamespace(env, namespaceId, tid, false);
  }
}
```

Reply:
I checked all our code for any references to `.call()` and the only time `Repo.call()` is seen is in `Fate.java` after `.isReady()`. The only other occurrence was in TraceRepo, whose `isReady` and `call` methods proxy the wrapped Repo. So, `.call()` is never run without first running `.isReady()`.

So, I think 1-pre-step solution is fine (and preferred, because it's simpler).

The only risk is running `mutateZooKeeper` multiple times, and I don't think that's a big deal. First, it could also happen if there was a failure in the 2-step solution so it doesn't change anything get rid of the second step. Second, we really want to make sure compactors get the message. The only other concern I can think of is whether `mutateZooKeeper` is idempotent. It should be, since that was already existing code in the CancelCompactions Repo.

There is one other thing I noticed about reusing the `CancelCompactions.mutateZooKeeper` code, that we may wish to revisit: that code seems to cancel currently running/scheduled compactions relative to some `flushID`. However, for delete, we really want to block all possible future compactions as well, so the implementation needed to cancel compactions for DeleteTable can just set the compaction ID to a fixed maximum value, rather than even bother checking what the current value is and mutating it.

Reply:
> So, I think 1-pre-step solution is fine (and preferred, because it's simpler).

yeah w/ your analysis that seems good.

> so the implementation needed to cancel compactions for DeleteTable can just set the compaction ID to a fixed maximum value, rather than even bother checking what the current value is and mutating it.

That sounds good, just need to ensure that overflow does not happen if some other does try to increment it.

Reply:
I'm about to push an update here based on the conversation. I think I captured everything except that I did not set the compaction ID to some maximum value as I did not see a case in the existing code where it checked to see if the currentValue +1 was greater than Long.MAX_VALUE. I wasn't sure what to do in this case.

Reply:
I think it's probably worth checking where all this field is incremented, to add that protection, and to use some max value (or non-numeric special string) to represent the deleting case. I can look into doing that as a follow on, if you don't want to include it in this PR.

Reply:
I would like to leave the compaction id changes to a different issue, mainly because I'm not comfortable enough with the code.

Reply:
Created follow-up issue for this in #2182